medmodels/data_synthesis/mtgan/modules/preprocessor.py

```diff
@@ -0,0 +1,590 @@
+"""Module for preprocessing the data for the MTGAN model.
+
+This module preprocesses the data to be used in the MTGAN model, following these steps:
+
+Steps:
+    1. Remove uncommon concepts.
+    2. Sample patients.
+    3. Find the first admission for each patient.
+    4. Remove admissions with less than min_codes_per_window codes.
+    5. Find the relative time windows for each patient per concept.
+
+It adds the following attributes to the MedRecord:
+    - first_admission_attribute: The first admission of the patient (first time they
+        have a connecting edge to a concept pertaining to the concepts_group).
+    - concept_index_attribute: The index of the concept. This is used to map the
+        concept to a numerical index that will be used in the model (connecting them
+        to the column on the boolean Data Matrix).
+    - absolute_time_window_attribute: The absolute time window for each edge. This is,
+        the number of days from the first admission divided by the time_interval_days.
+    - number_windows_attribute: The number of unique time windows for each patient.
+    - time_window_attribute: The relative time window for each edge. This means
+        collapsing the absolute time window to a ranking of unique time windows. For
+        instance, if a patient has absolute time windows [0, 2, 2, 3], the method will
+        assign relative time windows [0, 1, 1, 2] to the corresponding edges.
+"""
+
+import datetime
+import random
+from math import floor
+from typing import Dict, List, Tuple, TypedDict
+
+import pandas as pd
+from torch import nn
+
+from medmodels.medrecord.medrecord import MedRecord
+from medmodels.medrecord.querying import edge, node
+from medmodels.medrecord.types import (
+    Group,
+    MedRecordAttribute,
+    MedRecordValue,
+    NodeIndex,
+)
+from medmodels.treatment_effect.temporal_analysis import find_reference_edge
+
+
+class PreprocessingHyperparameters(TypedDict, total=True):
+    minimum_occurrences_concept: int
+    time_interval_days: int
+    minimum_concepts_per_window: int
+    number_of_sampled_patients: int
+
+
+class PreprocessingHyperparametersOptional(TypedDict, total=False):
+    minimum_occurrences_concept: int
+    time_interval_days: int
+    minimum_concepts_per_window: int
+    number_of_sampled_patients: int
+
+
+class PreprocessingAttributes(TypedDict):
+    first_admission_attribute: MedRecordAttribute
+    time_window_attribute: MedRecordAttribute
+    concept_index_attribute: MedRecordAttribute
+    number_of_windows_attribute: MedRecordAttribute
+    absolute_time_window_attribute: MedRecordAttribute
```

Comment:
missing attribute from the interface file. Intentional?

Reply:
Yes, realized there was no need for two attributes with the same functionality having different names.